### PR TITLE
[benchmark] Robust Measurements 

### PR DESCRIPTION
**Pull request by palimondo**
With the faster benchmarks in Swift Benchmark Suite (now that the Legacy Factor refactoring is finished), we don’t need to be so extremely frugal with the sample count. Gathering just the first 3 to 10 samples per benchmark was not very representative from the statistical point of view. There was a not insignificant chance that the benchmark did not yet reach a steady state. I suspect it was the reason for sporadic false changes reported from CI. (For most recent example see #26455.)

This PR adjusts the measurement method (in both the `Benchmark_Driver` as well as `run_smoke_bench`) to sample each benchmark for 250 ms and gather at minimum 10 samples. The `run_smoke_bench` now gathers up to 10 independent runs for the suspected changes. Ten independent runs with larger sample counts are enough to get statistically sound results, so it no longer spins in a measurement loop waiting for 10 no-change runs in a row.

### CODE REVIEW

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please smoke test

**Comment by palimondo** [comment]
@eeckstein Hmm… I should have started full test, to also exercise the python unit tests. But if you trust me it all passes locally, we could check this (assuming smoke test passes) based on my honor word, to help Joe get less erratic results ASAP.

**Comment by palimondo** [comment]
Hmm… I must have messed something up. Will check after returning from dog walk.

**Comment by palimondo** [comment]
Interesting… failure in 2nd iteration of Onone in the `ObjectiveCBridgeASCIIStringFromFile` benchmark. @Catfish-Man?

**Comment by palimondo** [comment]
@Catfish-Man, disregard. The error is in python code for merging results and has nothing with the particular benchmark.

**Comment by Catfish-Man** [comment]
If we get more stable benchmark results can we decrease the 7% reporting threshold?

**Comment by palimondo** [comment]
I can't figure out how am I sporadically reporting removed benchmarks whose runtime is 0 and how can they be missing in results dictionary during merge. Here it's too late now, I'll try again tomorrow. Sorry!

**Comment by palimondo** [comment]
This is some kind of a heisenbug… I'm going for vacation next week, so I will not be able to work on this. Maybe I'll push the little cleanup work I did in the meantime and see if that fixes the issue.

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci please test

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci please test

**Comment by palimondo** [comment]
@swift-ci please test

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci please test

**Comment by palimondo** [comment]
@swift-ci please test

**Comment by palimondo** [comment]
This isn’t really better… 🙁

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
I’ve backed out the finer threshold. It was too high of a goal to get working correctly at this moment.

@atrick, @jckarter mentioned to me that you had your own analysis of the sporadic false reports, I’d appreciate if you could share your perspective here.

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
I think this looks OK, but let me run one more benchmark.

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci test

**Comment by palimondo** [comment]
@swift-ci test

**Comment by palimondo** [comment]
Well, benchmarks with runtimes under 20 μs need to be adjusted anyway… and `Flatten` has its issues and also a pending fix, but that PR needs a bit of a refresh: #20552… so I’d argue this is fine?

**Comment by atrick** [comment]
Since my comments were alluded to above, I want to avoid confusion about the different kinds of variation we see in performance results. Performance is unpredictable in response to either changes to the compiler or to the inputs. I've made comments about these "optimizer instability" problems, or more generally performance unpredictability of the language. It's a problem because fundamentally good changes often produce bad results. The tradeoffs involved have to do with language features, peak benchmark performance, and time spent developing robust optimizations.

Run-to-run variation is different. It has to do with the design of the benchmarks and measurement methodology. I have not made any comments about this recently. The tradeoffs here are with turn around time, machine resources, precision and sensitivity, applicability to larger benchmarks, etc. To the extent that anyone pays attention to the benchmark scores, I do think it's important to fix "measurement instability". We just need to acknowledge that even with a perfectly deterministic benchmarking methodology, changes in benchmark scores across compiler changes can still be very misleading.

**Comment by palimondo** [comment]
@eeckstein please review

**Comment by eeckstein** [comment]
@palimondo I'll review this tomorrow. Please wait with merging until I get a chance to take a look.

**Comment by palimondo** [comment]
Sure. I’m back from vacation, so I hope to get to a better solution for the ensuring more stable result than the last 3 commits (which reverted back to the 7% threshold and 16x loop), before you get to do the review.

**Review by eeckstein — CHANGES_REQUESTED** [review]
FYI, when I tested the changes locally, I still saw some significant noise in some benchmarks.
This PR might improve the situation a bit, but it does not give us reliable "Robust Measurements" yet.

**Comment by palimondo** [comment]
@eeckstein
> but it does not give us reliable "Robust Measurements" yet.

Agreed. This needs more work. I'll address your other comments inline.

File: benchmark/scripts/run_smoke_bench
Comment (palimondo) [reply benchmark/scripts/run_smoke_bench]: It is not a typo. I've intended to gather samples for a quarter of a second to ensure benchmarks have reached steady state. From PR description:
> This PR adjusts the measurement method (in both the Benchmark_Driver as well as run_smoke_bench) to sample each benchmark for 250 ms and gather at minimum 10 samples. The run_smoke_bench now gathers up to 10 independent runs for the suspected changes. Ten independent runs with larger sample counts are enough to get statistically sound results, so it no longer spins in a measurement loop waiting for 10 no-change runs in a row.

Of course it makes the measurement take longer than before, because we are gathering more samples. The 2500 μs you've used originally were a reasonable trade-off at time when measuring the whole benchmark suite took much longer. But the proper baseline to compare this is before all the **legacy factor** refactoring. These have massively shortened the time to run the whole SBS. The point here is to use the time we've won and invest it into better quality of measurement: more samples. This way we can rely on better statistical methods in our analysis.

Here are my local timings per one full run of SBS with in various cases:

options | `O` | `Osize` | `Onone`
--|--|--|--
`--sample-time=0.0025 --num-samples=3` | 0:10 | 0:12 | 1:31
`--sample-time=0.0025 --num-samples=10` | 0:31 | 0:39 | 4:43
`--sample-time=0.25 --min-samples=10`<br>` --num-iters=1` | 1:31 | 2:07 | 6:12

There were 758 benchmarks in the pre-commit category of the SBS when I measured this locally, which — at 0.25 s per benchmark — should nominally take **3:10** for one full run of SBS.  Comparing this to the last line with my proposed measurement parameters, `O` and `Osize` take less time, because measurements are capped at 200 samples and end earlier for fast benchmarks. The `Onone` takes more than that because on my slow machine it takes longer to gather at least 10 samples per benchmark (that's also visible in the `--num-samples=10` case, which be the worst case scenario per iteration in our current system).

Your modern machines are much faster. Looking at the [console of the benchmark run](https://ci.swift.org/job/swift-PR-osx-perf/1156/console) in this PR and comparing it to [ numbers from another benchmark run](https://ci.swift.org/job/swift-PR-osx-smoke-perf/924/console) of the current `run_smoke_bench` from PR 26630 yields following times for single 0 iteration of SBS:

options | `O` | `Osize` | `Onone`
--|--|--|--
`--sample-time=0.0025 --num-samples=3` | 0:06 | 0:07 | 0:31
`--sample-time=0.25 --min-samples=10`<br>` --num-iters=1` | 0:34 | 0:39 | 2:21

The total time to run smoke benchmark was 3:19 with current method and 8:09 with the one in this PR, a slowdown of approximately 2.5. Note that the time from initial git pull to finished compilation on this PR was 19:23.

File: benchmark/scripts/Benchmark_Driver
Comment (palimondo) [reply benchmark/scripts/Benchmark_Driver]: This is the original `Benchmark_Driver`, which used default sample time of 1 second up until now. This is a 4x speedup of its measurement throughput. From the commit description in b4c8153:
> With the faster benchmarks in the Swift Benchmark Suite, we now don’t need to spend a whole 1 second measuring each one of them. So I’m adjusting the Benchmark_Driver to sample each one for 250 ms. Quarter of a second should be plenty of time to reach a stable state in most cases.
>
> For the slow ones and measuring of the unoptimized Onone build, let’s improve the quality of statistics by gathering at least 10 samples.

File: benchmark/scripts/run_smoke_bench
Comment (palimondo) [reply benchmark/scripts/run_smoke_bench]: The last 3 commits here were done from mobile and were my attempt to somehow address the apparent instability by reverting to the old brute force methods. I'm pretty sure that my original approach to stop at 10 measurements was a correct one, but I need to handle the special cases of benchmarks with runtimes under 20 μs as a special case in result comparison and also adding filtering for min outliers like is the case with `ObjectiveCBridgeStubToNSDateRef`.

Comment (palimondo) [reply benchmark/scripts/run_smoke_bench]: IMG_0747

Comment (palimondo) [reply benchmark/scripts/run_smoke_bench]: Another example of benchmark that jumped up here during the first run, which seems like something that has about 2:10 chance to look unstable is we stick to the naive comparison of MIN values is `DataAppenDataSmallToSmall`: 
IMG_0745

Comment (palimondo) [reply benchmark/scripts/run_smoke_bench]: I’d like to fix that issue by expanding on the dubious indicator **(?)** in the ratio column, so that it revealed more details about the underlying sample distribution, to enable us to make more informed decisions in such cases.

File: benchmark/scripts/run_smoke_bench
Comment (palimondo) [reply benchmark/scripts/run_smoke_bench]: Note that going back to capping the iterations at 10 (after addressing the “instabilities”), might slightly improve the total runtime of this method, to maybe somewhere around 2x slowdown.

Comment (palimondo) [reply benchmark/scripts/run_smoke_bench]: I will also look at running benchmarks in parallel with this method, which will speed this back up again, depending on the number of cores on the machine.

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
This is still work in progress. I have modified the benchmark reporting for dubious result comparisons to print out empirical sample distribution ([ventiles](https://en.wiktionary.org/wiki/ventile)), to get better insight into the nature of problematic benchmarks. In the current process, the reported changes have samples gathered from 10 independent measurements. These are merged together the form a composite sample distribution.

In my local measurements, there is a class of benchmarks that tend to show up in the reports as false changes, given that I am comparing the same binary against itself. I'm interested to see to what extent are these HW specific and to what extent are these _"unstable"_ benchmarks platform independent.

Next step would be to devise a heuristic, other than comparing just MIN values between NEW and OLD, that deals with the perceived instability caused by this naive approach.

File: benchmark/scripts/test_compare_perf_tests.py
Comment (palimondo) [inline benchmark/scripts/test_compare_perf_tests.py]: This is a visualization of an example of the problematic benchmark used in these unit tests:
Dubious Benchmark

Comment (palimondo) [reply benchmark/scripts/test_compare_perf_tests.py]: These are two independent executions, picked to showcase the rarely occurring but very significant variation present in one of them.

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
There are clearly a few benchmarks with too small runtime, some of them because of wrong `legacyFactor`s due to the large performance gap between my machine and the CI machines, so I’m running a full `BenchmarkDoctor` check on CI to get a list of those that need to be re-adjusted.

**Comment by palimondo** [comment]
Hah, I had to pull the benchmark check from the console log, because I’ve missed a conditional to print it in HTML. But I got that list, don’t worry.

**Comment by palimondo** [comment]
The purpose of this PR is to find more statistically rigorous method of detecting performance changes in our benchmarks, while minimizing [type I and type II errors](https://en.wikipedia.org/wiki/Type_I_and_type_II_errors). The altered measurement technique has apparently increased the chance of type I errors — false positives — because it has eliminated the averaging of multiple samples and has significantly increased the number of samples gathered.

From the ventiles collected from the altered dubious reports (they are marked by **(?)** in the Ratio column) I’ve charted the runtimes to see what is the empirical distribution of the samples from benchmarks that are popping up with this more sensitive measurement technique.

## Flatten
Most notorious are the Flatten benchmarks, which sporadically appear even with the current measurement process:
IMG_0874
IMG_0875

These clearly demonstrate that for certain benchmarks, using the [minimum as a location estimator is genuinely problematic](https://palimondo.github.io/robust-microbench/detecting-changes.html#issues-with-minimum-as-location-estimator).

We have to remember that the samples reported in tables and visualized in the charts above are aggregates from 10 independent measurement runs. But here we don’t have all the samples, just the minimum and [ventiles](https://en.wiktionary.org/wiki/ventile#English), so the lower end of the distribution is also over-represented (10x MIN and 5th percentiles). It is not clear if the extreme outliers appear rarely in every measurement or if they come from one measurement with different distribution.

We should probably fix these benchmarks by lowering their runtime anyway (#20552), but they present interesting challenge, which we could try to solve by applying more sophisticated statistical techniques, than just looking if there’s more than 5% difference between the minimum values.

## ObjectiveCBridgeStubToNSDateRef
This benchmarks has slightly different empirical distribution between the `O` and `Onone`:
IMG_0876
IMG_0877

Here again, minimum proves to be poor location estimate for this benchmark, because it appears to have the highest variance, when compared to other ventiles. Interestingly, the sample distribution of this benchmarks is also quite wide, with [interquartile range](https://en.wikipedia.org/wiki/Interquartile_range) 12-14%.

## Others
Most other falsely reported benchmarks show that the minimum can be noisy location indicator, compared to the 5th or 10th percentile.

(To be continued...)

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by palimondo** [comment]
> Can one of the admins verify this patch?

I’m still working on this. Just got bogged down in a job that actually pays the bills… 😐

**Comment by palimondo** [comment]
My current plan to make this measurement method more robust:
1. When merging `PerformanceTestResult`s, keep the original `PerformanceTestSample`s from all independent runs. 
2. These will be used to choose the most stable (least variable) location estimate for the `ResultComparison`: compute variance for MIN, V1, V2, Q1, MED (these are 0th, 5th, 10th, 25th a 50th percentiles respectively) for OLD and NEW result (these will be composed from at least 3 and up to 10 independent runs in the `run_smoke_benchmark`), multiply these<sup>*</sup> with each other and pick the one with lowest value as the most stable location estimate for the sample distributions. This will then be used in the comparison as usual.
3. Adjust the `is_dubious` analysis on `ResultComparison` to be a bit more sophisticated than the current range overlap between the OLD and NEW samples. Keep the ventiles printout that shows the cumulative empirical sample distribution to aid in decision making should such case arise.

<sup>*</sup> *Since variance is standard deviation squared, using a simple multiplication as formula to pick the most stable location should work fine, thanks to its exponential nature… But I’ll try to simulate and visualize this with few plots in R tomorrow, before implementing this strategy in Python.*

CC @eeckstein @gottesmm 
Please don’t push forward with #30085 yet, temporarily reverting #29719 to land this is complicated enough. I have sunk a lot of thought and work into this PR and think it will be a major improvement to our measurement methodology. I’ll finish this over the weekend and will help with the Python 3 conversion thereafter.

**Comment by gottesmm** [comment]
@palimondo if you are going to do this, my suggestion is to do this in a way that you can land this without flipping the final switch. That way we do not rush this in and all of a sudden @swift-ci benchmark changes all of a sudden in a rushed way. I am concerned with putting things into the tree just to get them into the tree.

By landing it we can make sure that the current approach still works, we can flip the switch gracefully when you have time to fully look into any problems that come up. I am assuming since you said that you need to use time to work your normal job that it may be difficult for you to spend more time immediately after landing this if any issues come up. Is my interpretation correct?

**Comment by palimondo** [comment]
@gottesmm 
> By landing it we can make sure that the current approach still works, we can flip the switch gracefully when you have time to fully look into any problems that come up. I am assuming since you said that you need to use time to work your normal job that it may be difficult for you to spend more time immediately after landing this if any issues come up. Is my interpretation correct?

I’ll keep the final “flipping the switch" as last few commits, so that we can separate them as necessary. But @eeckstein’s reviews and merges usually happen in my after work hours anyway, so my evenings are all yours if the need to fix anything will arise. (The bill paying job was just a reference to my dashed dreams of turning the work on Swift into paying remote job. Apple’s hiring policy that requires relocation even for work on distributed open source project is just frustrating… 😞)

**Comment by gottesmm** [comment]
@palimondo thanks. As long as we can revert easily flipping the switch, I am 100% ok! Make sure to separate your changes out carefully!

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
OK, there is probably still some fine tuning work ahead of us, but let’s see if the basic idea is correct. I’ll run a few benchmarks to gather more data.

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
This is a bit unexpected, but I see this also in my local testing. Let me dump quartiles for all independent runs at once and I’ll do some analysis tomorrow.

**Comment by eeckstein** [comment]
@palimondo To reiterate: increasing the sample time in run _smoke_bench by 100x is not acceptable. I cannot approve this PR with this change.

**Comment by palimondo** [comment]
@eeckstein I understand your point. Let’s make this more manageable:
* I have split off the groundwork for robust measurements to it own PR #30158. It was the bulk of the work here and the part I really don’t want to loose, because of all the conflicts coming from automatic formatting changes from #29719. I want to land this ASAP.
* We can discuss the measurement parameters and the merits of proposed changes from this PR later. It will be much smaller change once the groundwork lands and it doesn’t have to block the Python 3 support and all the automatic formatting changes. 
* I’d like to hear more of your thoughts about that, since #29719 has landed without any justification in its description. Was there some plan discussed in Swift forums that I missed? I’m assuming @Rostepher’s work on that will can be reproduced by some automated step (missing docs?)?

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by Rostepher** [comment]
@palimondo The goal is to add another Python validation-test to run `utils/python_format.py --check` which should enforce the black style on all our Python sources. Details are in #29701.

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by gottesmm** [comment]
I don't understand this new output and it definitely needs to be behind a detail drop down.

**Comment by palimondo** [comment]
@gottesmm I’m sorry for spamming you all with these verbose reports. I needed these just temporarily to gather more data and devise better strategy for coping with the issue of false changes. Now it’s clear the extreme outliers are coming from “lucky runs”, which then skew the aggregated sample distribution, used for the final comparison of OLD and NEW. The selection of most stable location from f888aefe9d989fee26fac6a2f8ef3d6deeb6c15c does not help with this problem, because the individual runs have the same shape of distribution, the lucky runs are just shifted in location and we keep comparing the MINs. I think I’ll need to detect and throw out such outlier runs. This explains why `run_smoke_bench` currently needs to go up to 20 or more iterations to get the “lucky run” from the other result and doesn’t always succeed.

**Comment by palimondo** [comment]
@swift-ci please python lint

**Comment by palimondo** [comment]
@swift-ci please python lint

**Comment by palimondo** [comment]
@swift-ci python lint

**Comment by palimondo** [comment]
@swift-ci python lint

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please smoke test

**Comment by palimondo** [comment]
@swift-ci please smoke test

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
Wow, now this is upsetting! Switching away from MIN as location estimate on these benchmarks did backfire tremendously. I have seen nothing like this locally. I need to re-enable that verbose dump to understand the distribution of individual runs for these cases.

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@shahmishal @eeckstein Did @swift-ci start to invoke `run_smoke_bench` with `--num-reruns` equal to 22 recently? I’m surprised by the number of independent iterations I’ve dumped. It seems we were using the default value just few days ago…

**Comment by CodaFi** [comment]
@palimondo This PR is enormous! The python formatting changes need to be broken out separately from the benchmark formatting changes.

**Comment by palimondo** [comment]
@CodaFi Python formatting was broken out already in a separate PR this one
depends on, but that's neither here nor there since #30085 has landed :-/
resolving the merge conflicts will be a slog. I'll get back to this,
eventually.

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**
